README.md
[![Crates.io](https://img.shields.io/crates/v/tuikit.svg)](https://crates.io/crates/tuikit) [![Build Status](https://github.com/lotabout/tuikit/workflows/Build%20&%20Test/badge.svg)](https://github.com/lotabout/tuikit/actions?query=workflow%3A%22Build+%26+Test%22)

## Tuikit

Tuikit is a TUI library for writing terminal UI applications. Highlights:

- Thread safe.
- Support non-fullscreen mode as well as fullscreen mode.
- Support `Alt` keys, mouse events, etc.
- Buffering for efficient rendering.

Tuikit is modeld after [termbox](https://github.com/nsf/termbox) which views the
terminal as a table of fixed-size cells and input being a stream of structured
messages.

**WARNING**: The library is not stable yet, the API might change.

## Usage

In your `Cargo.toml` add the following:

```toml
[dependencies]
tuikit = "*"
```

And if you'd like to use the latest snapshot version:

```toml
[dependencies]
tuikit = { git = "https://github.com/lotabout/tuikit.git" }
```

Here is an example (could also be run by `cargo run --example hello-world`):

```rust
use tuikit::prelude::*;
use std::cmp::{min, max};

fn main() {
    let term: Term<()> = Term::with_height(TermHeight::Percent(30)).unwrap();
    let mut row = 1;
    let mut col = 0;

    let _ = term.print(0, 0, "press arrow key to move the text, (q) to quit");
    let _ = term.present();

    while let Ok(ev) = term.poll_event() {
        let _ = term.clear();
        let _ = term.print(0, 0, "press arrow key to move the text, (q) to quit");

        let (width, height) = term.term_size().unwrap();
        match ev {
            Event::Key(Key::ESC) | Event::Key(Key::Char('q')) => break,
            Event::Key(Key::Up) => row = max(row-1, 1),
            Event::Key(Key::Down) => row = min(row+1, height-1),
            Event::Key(Key::Left) => col = max(col, 1)-1,
            Event::Key(Key::Right) => col = min(col+1, width-1),
            _ => {}
        }

        let attr = Attr{ fg: Color::RED, ..Attr::default() };
        let _ = term.print_with_attr(row, col, "Hello World! 你好！今日は。", attr);
        let _ = term.set_cursor(row, col);
        let _ = term.present();
    }
}
```

## Layout

`tuikit` provides `HSplit`, `VSplit` and `Win` for managing layouts:

1. `HSplit` allow you to split area horizontally into pieces.
2. `VSplit` works just like `HSplit` but splits vertically.
3. `Win` do not split, it could have margin, padding and border.

For example:

```rust
use tuikit::prelude::*;

struct Model(String);

impl Draw for Model {
    fn draw(&self, canvas: &mut dyn Canvas) -> Result<()> {
        let (width, height) = canvas.size()?;
        let message_width = self.0.len();
        let left = (width - message_width) / 2;
        let top = height / 2;
        let _ = canvas.print(top, left, &self.0);
        Ok(())
    }
}

impl Widget for Model{}

fn main() {
    let term: Term<()> = Term::with_height(TermHeight::Percent(50)).unwrap();
    let model = Model("middle!".to_string());

    while let Ok(ev) = term.poll_event() {
        if let Event::Key(Key::Char('q')) = ev {
            break;
        }
        let _ = term.print(0, 0, "press 'q' to exit");

        let hsplit = HSplit::default()
            .split(
                VSplit::default()
                    .basis(Size::Percent(30))
                    .split(Win::new(&model).border(true).basis(Size::Percent(30)))
                    .split(Win::new(&model).border(true).basis(Size::Percent(30)))
            )
            .split(Win::new(&model).border(true));

        let _ = term.draw(&hsplit);
        let _ = term.present();
    }
}
```

The split algorithm is simple:

1. Both `HSplit` and `VSplit` will take several `Split` where a `Split` would
   contains:
    1. basis, the original size
    2. grow, the factor to grow if there is still enough room
    3. shrink, the factor to shrink if there is not enough room
2. `HSplit/VSplit` will count the total width/height(basis) of the split items
3. Judge if the current width/height is enough or not for the split items
4. shrink/grow the split items according to their grow/shrink: `factor / sum(factors)`
5. If still not enough room, the last one(s) would be set width/height 0

## References

`Tuikit` borrows ideas from lots of other projects:

- [rustyline](https://github.com/kkawakam/rustyline) Readline Implementation in Rust.
    - How to enter the raw mode.
    - Part of the keycode parsing logic.
- [termion](https://gitlab.redox-os.org/redox-os/termion) A bindless library for controlling terminals/TTY.
    - How to parse mouse events.
    - How to enter raw mode.
- [rustbox](https://github.com/gchp/rustbox) and [termbox](https://github.com/nsf/termbox)
    - The idea of viewing terminal as table of fixed cells.
- [termfest](https://github.com/agatan/termfest) Easy TUI library written in Rust
    - The buffering idea.


CHANGELOG.md
# Change Logs

## v0.4.5: 2021-02-15

Feature:
- Travis CI -> Github Actions

Fix:
- parse missing keynames(ctrl-up/down/left/right)

## v0.4.4: 2021-02-14

Feature:
- tuikit now returns concrete errors

Fix:
- restore the `clear_on_exit` behavior
- key listener no longer quit(hang) on unknown sequence

## v0.4.3: 2021-01-03

Feature:
- support bracketed paste mode

## v0.4.2: 2020-10-20

Fix:
- click/wheel events' row were not adjusted in non-fullscreen mode

## v0.4.1: 2020-10-18

Fix:
- `Term` not paused on drop.

## v0.4.0: 2020-10-15

Feature:
- support `hold` option that don't start term on creation.
- support user defined event.
- unify result types

## v0.3.4: 2020-10-06

Feature:
- widget `win` support header and right prompt
- new widget: `stack` for stacking widget bottom up
- keyboard now parses double click events
    - in this mode, `MousePress` event would no longer be generated
- keyboard now merges consecutive wheel events

Fix:
- show cursor when quiting alternate screen

## v0.3.3: 2020-06-26

- fix [skim#308](https://github.com/lotabout/skim/issues/308): skim hang on
    initialization

## v0.3.2: 2020-04-01

- fix skim#259 release lock correctly on pause
- fix skim#277: x10 mouse event was capped

## v0.3.1: 2020-02-05

- fix skim#232: use `cfmakeraw` to enable raw mode
- fix build with rust 1.32.0

## v0.3.0: 2020-01-30

Feature:
- Feature: option to clear screen or not after exit.
- Feature: new trait `Widget`

Bug fixes:
- fix skim#255: parse `space` as key ` `
- reset mouse status before exit.
- fix: adjust mouse position(row)'s origin

Examples:
- 256color_on_screen: reset attributes before flush
- fix #10: output help in split example
- get_keys: disable mouse before existing
- all: make examples quit on Ctrl-C

Depedency Update:
- `term` to `0.6`

## v0.2.9: 2019-07-28

Fix: [skim#192](https://github.com/lotabout/skim/issues/192): Start drawing in
a clean line.

## v0.2.8: 2019-06-05

Update dependenncy `nix` to `0.14`.

## v0.2.7: 2019-06-04

Features:
- Implement `From` trait for variaous struct
    * `From<Color> for Attr`
    * `From<Effect> for Attr`
    * `From<char> for Cell`
- `win/split` now accept `Into<...>` struct. Previously when initializing
    splits, you need to write `split.basis(10.into())`,
    now it's just `split.basis(10)`.
- Implement builder pattern for `Attr`. We could now do
    `Attr::default().fg(...).bg(...)`.
- Add two user defined event(`User1` and `User2`). Use it for your own need.

Bug fixes:
- fix compilation error on FreeBSD.

## v0.2.6: 2019-03-28

Reduce CPU usage on idle.

## v0.2.5: 2019-03-28

Clear screen on resize

## v0.2.4: 2019-03-22

Fix: ESC key not working

## v0.2.3: 2019-03-23

- Support more alt keys
- impl `Draw` for `Box<T: Draw>`

## v0.2.2: 2019-03-19

API change: `Draw::content_size` -> `Draw::size_hint` and returns
`Option<usize>`. So that `None` could indicates "I don't know".

## v0.2.1: 2019-03-17

- fix: build failed with rust 2018 (1.31.0)

## v0.2.0: 2019-03-17

Feature:
- Support layout(e.g. `HSplit`, `VSplit`)
- `term.send_event` to inject event to `Term`'s event loop
- `use tuikit::prelude::*` to simplify import

## v0.1.5: 2019-03-02

Fix: Synchronize the pause and restart event.

## v0.1.4: 2019-02-25

Fix: output will replace raw ESC(`\x1b`) with `?` so that terminal won't mess up.

## v0.1.3: 2019-02-24

Fix: report cursor position (0, 0) on terminals that doesn't support CPR.

## v0.1.2: 2019-02-24

Features:
- support specifying `min-height` and `max-height`
- screen: able to iterate over all cells
- attr: add `extend` method for composing `Attr`.

Bug Fixes:
- #1 Increase timeout(to 300ms) on initialize to support slow terminals
- #3 erase contents on exit
- screen: fix panic on height/width of `0`
- fix some key parsing error


src/input.rs
//! module to handle keystrokes
//!
//! ```no_run
//! use tuikit::input::KeyBoard;
//! use tuikit::key::Key;
//! use std::time::Duration;
//! let mut keyboard = KeyBoard::new_with_tty();
//! let key = keyboard.next_key();
//! ```

use std::fs::File;
use std::io::prelude::*;
use std::os::unix::io::AsRawFd;
use std::os::unix::io::FromRawFd;
use std::sync::Arc;
use std::time::{Duration, Instant};

use nix::fcntl::{fcntl, FcntlArg, OFlag};

use crate::error::TuikitError;
use crate::key::Key::*;
use crate::key::{Key, MouseButton};
use crate::raw::get_tty;
use crate::spinlock::SpinLock;
use crate::sys::file::wait_until_ready;
use crate::Result;

pub trait ReadAndAsRawFd: Read + AsRawFd + Send {}

const KEY_WAIT: Duration = Duration::from_millis(10);
const DOUBLE_CLICK_DURATION: u128 = 300;

impl<T> ReadAndAsRawFd for T where T: Read + AsRawFd + Send {}

pub struct KeyBoard {
    file: Box<dyn ReadAndAsRawFd>,
    sig_tx: Arc<SpinLock<File>>,
    sig_rx: File,
    // bytes will be poped from front, normally the buffer size will be small(< 10 bytes)
    byte_buf: Vec<u8>,

    raw_mouse: bool,
    next_key: Option<Result<Key>>,
    last_click: Key,
    last_click_time: SpinLock<Instant>,
}

// https://www.xfree86.org/4.8.0/ctlseqs.html
// http://man7.org/linux/man-pages/man4/console_codes.4.html
impl KeyBoard {
    pub fn new(file: Box<dyn ReadAndAsRawFd>) -> Self {
        // the self-pipe trick for interrupt `select`
        let (rx, tx) = nix::unistd::pipe().expect("failed to set pipe");

        // set the signal pipe to non-blocking mode
        let flag = fcntl(rx, FcntlArg::F_GETFL).expect("Get fcntl failed");
        let mut flag = OFlag::from_bits_truncate(flag);
        flag.insert(OFlag::O_NONBLOCK);
        let _ = fcntl(rx, FcntlArg::F_SETFL(flag));

        // set file to non-blocking mode
        let flag = fcntl(file.as_raw_fd(), FcntlArg::F_GETFL).expect("Get fcntl failed");
        let mut flag = OFlag::from_bits_truncate(flag);
        flag.insert(OFlag::O_NONBLOCK);
        let _ = fcntl(file.as_raw_fd(), FcntlArg::F_SETFL(flag));

        KeyBoard {
            file,
            sig_tx: Arc::new(SpinLock::new(unsafe { File::from_raw_fd(tx) })),
            sig_rx: unsafe { File::from_raw_fd(rx) },
            byte_buf: Vec::new(),
            raw_mouse: false,
            next_key: None,
            last_click: Key::Null,
            last_click_time: SpinLock::new(Instant::now()),
        }
    }

    pub fn new_with_tty() -> Self {
        Self::new(Box::new(
            get_tty().expect("KeyBoard::new_with_tty: failed to get tty"),
        ))
    }

    pub fn raw_mouse(mut self, raw_mouse: bool) -> Self {
        self.raw_mouse = raw_mouse;
        self
    }

    pub fn get_interrupt_handler(&self) -> KeyboardHandler {
        KeyboardHandler {
            handler: self.sig_tx.clone(),
        }
    }

    fn fetch_bytes(&mut self, timeout: Duration) -> Result<()> {
        let mut reader_buf = [0; 1];

        // clear interrupt signal
        while let Ok(_) = self.sig_rx.read(&mut reader_buf) {}

        wait_until_ready(
            self.file.as_raw_fd(),
            Some(self.sig_rx.as_raw_fd()),
            timeout,
        )?; // wait timeout

        self.read_unread_bytes();
        Ok(())
    }

    fn read_unread_bytes(&mut self) {
        let mut reader_buf = [0; 1];
        while let Ok(_) = self.file.read(&mut reader_buf) {
            self.byte_buf.push(reader_buf[0]);
        }
    }

    #[allow(dead_code)]
    fn next_byte(&mut self) -> Result<u8> {
        self.next_byte_timeout(Duration::new(0, 0))
    }

    fn next_byte_timeout(&mut self, timeout: Duration) -> Result<u8> {
        trace!("next_byte_timeout: timeout: {:?}", timeout);
        if self.byte_buf.is_empty() {
            self.fetch_bytes(timeout)?;
        }

        trace!("next_byte_timeout: after fetch, buf = {:?}", self.byte_buf);
        Ok(self.byte_buf.remove(0))
    }

    #[allow(dead_code)]
    fn next_char(&mut self) -> Result<char> {
        self.next_char_timeout(Duration::new(0, 0))
    }

    fn next_char_timeout(&mut self, timeout: Duration) -> Result<char> {
        trace!("next_char_timeout: timeout: {:?}", timeout);
        if self.byte_buf.is_empty() {
            self.fetch_bytes(timeout)?;
        }

        trace!("get_chars: buf: {:?}", self.byte_buf);
        let bytes = std::mem::replace(&mut self.byte_buf, Vec::new());
        match String::from_utf8(bytes) {
            Ok(string) => {
                let ret = string
                    .chars()
                    .next()
                    .expect("failed to get next char from input");
                self.byte_buf
                    .extend_from_slice(&string.as_bytes()[ret.len_utf8()..]);
                Ok(ret)
            }
            Err(error) => {
                let valid_up_to = error.utf8_error().valid_up_to();
                let bytes = error.into_bytes();
                let string = String::from_utf8_lossy(&bytes[..valid_up_to]);
                let ret = string
                    .chars()
                    .next()
                    .expect("failed to get next char from input");
                self.byte_buf.extend_from_slice(&bytes[ret.len_utf8()..]);
                Ok(ret)
            }
        }
    }

    fn merge_wheel(&mut self, current_key: Result<Key>) -> (Result<Key>, Option<Result<Key>>) {
        match current_key {
            Ok(Key::MousePress(key @ MouseButton::WheelUp, row, col))
            | Ok(Key::MousePress(key @ MouseButton::WheelDown, row, col)) => {
                let mut count = 1;
                let mut o_next_key;
                loop {
                    o_next_key = self.try_next_raw_key();
                    match o_next_key {
                        Some(Ok(Key::MousePress(k, r, c))) if key == k && row == r && col == c => {
                            count += 1
                        }
                        _ => break,
                    }
                }

                match key {
                    MouseButton::WheelUp => (Ok(Key::WheelUp(row, col, count)), o_next_key),
                    MouseButton::WheelDown => (Ok(Key::WheelDown(row, col, count)), o_next_key),
                    _ => unreachable!(),
                }
            }
            _ => (current_key, None),
        }
    }

    pub fn next_key(&mut self) -> Result<Key> {
        self.next_key_timeout(Duration::new(0, 0))
    }

    pub fn next_key_timeout(&mut self, timeout: Duration) -> Result<Key> {
        if self.raw_mouse {
            return self.next_raw_key_timeout(timeout);
        }

        let next_key = if self.next_key.is_some() {
            self.next_key.take().unwrap()
        } else {
            // fetch next key
            let next_key = self.next_raw_key_timeout(timeout);
            let (next_key, next_next_key) = self.merge_wheel(next_key);
            self.next_key = next_next_key;
            next_key
        };

        // parse double click
        match next_key {
            Ok(key @ MousePress(..)) => {
                if let MousePress(button, row, col) = key {
                    let ret = if key == self.last_click
                        && self.last_click_time.lock().elapsed().as_millis() < DOUBLE_CLICK_DURATION
                    {
                        DoubleClick(button, row, col)
                    } else {
                        self.last_click = key;
                        SingleClick(button, row, col)
                    };

                    *self.last_click_time.lock() = Instant::now();
                    Ok(ret)
                } else {
                    unreachable!();
                }
            }
            _ => return next_key,
        }
    }

    #[allow(dead_code)]
    fn next_raw_key(&mut self) -> Result<Key> {
        self.next_raw_key_timeout(Duration::new(0, 0))
    }

    fn try_next_raw_key(&mut self) -> Option<Result<Key>> {
        match self.next_raw_key_timeout(KEY_WAIT) {
            Ok(key) => Some(Ok(key)),
            Err(TuikitError::Timeout(_)) => None,
            Err(error) => Some(Err(error)),
        }
    }

    /// Wait `timeout` until next key stroke
    fn next_raw_key_timeout(&mut self, timeout: Duration) -> Result<Key> {
        trace!("next_raw_key_timeout: {:?}", timeout);
        let ch = self.next_char_timeout(timeout)?;
        match ch {
            '\u{00}' => Ok(Ctrl(' ')),
            '\u{01}' => Ok(Ctrl('a')),
            '\u{02}' => Ok(Ctrl('b')),
            '\u{03}' => Ok(Ctrl('c')),
            '\u{04}' => Ok(Ctrl('d')),
            '\u{05}' => Ok(Ctrl('e')),
            '\u{06}' => Ok(Ctrl('f')),
            '\u{07}' => Ok(Ctrl('g')),
            '\u{08}' => Ok(Ctrl('h')),
            '\u{09}' => Ok(Tab),
            '\u{0A}' => Ok(Ctrl('j')),
            '\u{0B}' => Ok(Ctrl('k')),
            '\u{0C}' => Ok(Ctrl('l')),
            '\u{0D}' => Ok(Enter),
            '\u{0E}' => Ok(Ctrl('n')),
            '\u{0F}' => Ok(Ctrl('o')),
            '\u{10}' => Ok(Ctrl('p')),
            '\u{11}' => Ok(Ctrl('q')),
            '\u{12}' => Ok(Ctrl('r')),
            '\u{13}' => Ok(Ctrl('s')),
            '\u{14}' => Ok(Ctrl('t')),
            '\u{15}' => Ok(Ctrl('u')),
            '\u{16}' => Ok(Ctrl('v')),
            '\u{17}' => Ok(Ctrl('w')),
            '\u{18}' => Ok(Ctrl('x')),
            '\u{19}' => Ok(Ctrl('y')),
            '\u{1A}' => Ok(Ctrl('z')),
            '\u{1B}' => self.escape_sequence(),
            '\u{7F}' => Ok(Backspace),
            ch => Ok(Char(ch)),
        }
    }

    fn escape_sequence(&mut self) -> Result<Key> {
        let seq1 = self.next_char_timeout(KEY_WAIT).unwrap_or('\u{1B}');
        match seq1 {
            '[' => self.escape_csi(),
            'O' => self.escape_o(),
            _ => self.parse_alt(seq1),
        }
    }

    fn parse_alt(&mut self, ch: char) -> Result<Key> {
        match ch {
            '\u{1B}' => {
                match self.next_byte_timeout(KEY_WAIT) {
                    Ok(b'[') => {}
                    Ok(c) => {
                        return Err(TuikitError::UnknownSequence(format!("ESC ESC {}", c)));
                    }
                    Err(_) => return Ok(ESC),
                }

                match self.escape_csi() {
                    Ok(Up) => Ok(AltUp),
                    Ok(Down) => Ok(AltDown),
                    Ok(Left) => Ok(AltLeft),
                    Ok(Right) => Ok(AltRight),
                    Ok(PageUp) => Ok(AltPageUp),
                    Ok(PageDown) => Ok(AltPageDown),
                    _ => Err(TuikitError::UnknownSequence(format!("ESC ESC [ ..."))),
                }
            }
            '\u{00}' => Ok(CtrlAlt(' ')),
            '\u{01}' => Ok(CtrlAlt('a')),
            '\u{02}' => Ok(CtrlAlt('b')),
            '\u{03}' => Ok(CtrlAlt('c')),
            '\u{04}' => Ok(CtrlAlt('d')),
            '\u{05}' => Ok(CtrlAlt('e')),
            '\u{06}' => Ok(CtrlAlt('f')),
            '\u{07}' => Ok(CtrlAlt('g')),
            '\u{08}' => Ok(CtrlAlt('h')),
            '\u{09}' => Ok(AltTab),
            '\u{0A}' => Ok(CtrlAlt('j')),
            '\u{0B}' => Ok(CtrlAlt('k')),
            '\u{0C}' => Ok(CtrlAlt('l')),
            '\u{0D}' => Ok(AltEnter),
            '\u{0E}' => Ok(CtrlAlt('n')),
            '\u{0F}' => Ok(CtrlAlt('o')),
            '\u{10}' => Ok(CtrlAlt('p')),
            '\u{11}' => Ok(CtrlAlt('q')),
            '\u{12}' => Ok(CtrlAlt('r')),
            '\u{13}' => Ok(CtrlAlt('s')),
            '\u{14}' => Ok(CtrlAlt('t')),
            '\u{15}' => Ok(CtrlAlt('u')),
            '\u{16}' => Ok(CtrlAlt('v')),
            '\u{17}' => Ok(CtrlAlt('w')),
            '\u{18}' => Ok(CtrlAlt('x')),
            '\u{19}' => Ok(AltBackTab),
            '\u{1A}' => Ok(CtrlAlt('z')),
            '\u{7F}' => Ok(AltBackspace),
            ch => Ok(Alt(ch)),
        }
    }

    fn escape_csi(&mut self) -> Result<Key> {
        let cursor_pos = self.parse_cursor_report();
        if cursor_pos.is_ok() {
            return cursor_pos;
        }

        let seq2 = self.next_byte_timeout(KEY_WAIT)?;
        match seq2 {
            b'0' | b'9' => Err(TuikitError::UnknownSequence(format!("ESC [ {:x?}", seq2))),
            b'1'..=b'8' => self.extended_escape(seq2),
            b'[' => {
                // Linux Console ESC [ [ _
                let seq3 = self.next_byte_timeout(KEY_WAIT)?;
                match seq3 {
                    b'A' => Ok(F(1)),
                    b'B' => Ok(F(2)),
                    b'C' => Ok(F(3)),
                    b'D' => Ok(F(4)),
                    b'E' => Ok(F(5)),
                    _ => Err(TuikitError::UnknownSequence(format!("ESC [ [ {:x?}", seq3))),
                }
            }
            b'A' => Ok(Up),    // kcuu1
            b'B' => Ok(Down),  // kcud1
            b'C' => Ok(Right), // kcuf1
            b'D' => Ok(Left),  // kcub1
            b'H' => Ok(Home),  // khome
            b'F' => Ok(End),
            b'Z' => Ok(BackTab),
            b'M' => {
                // X10 emulation mouse encoding: ESC [ M Bxy (6 characters only)
                let cb = self.next_byte_timeout(KEY_WAIT)?;
                // (1, 1) are the coords for upper left.
                let cx = self.next_byte_timeout(KEY_WAIT)?.saturating_sub(32) as u16 - 1; // 0 based
                let cy = self.next_byte_timeout(KEY_WAIT)?.saturating_sub(32) as u16 - 1; // 0 based
                match cb & 0b11 {
                    0 => {
                        if cb & 0x40 != 0 {
                            Ok(MousePress(MouseButton::WheelUp, cy, cx))
                        } else {
                            Ok(MousePress(MouseButton::Left, cy, cx))
                        }
                    }
                    1 => {
                        if cb & 0x40 != 0 {
                            Ok(MousePress(MouseButton::WheelDown, cy, cx))
                        } else {
                            Ok(MousePress(MouseButton::Middle, cy, cx))
                        }
                    }
                    2 => Ok(MousePress(MouseButton::Right, cy, cx)),
                    3 => Ok(MouseRelease(cy, cx)),
                    _ => Err(TuikitError::UnknownSequence(format!(
                        "ESC M {:?}{:?}{:?}",
                        cb, cx, cy
                    ))),
                }
            }
            b'<' => {
                // xterm mouse encoding:
                // ESC [ < Cb ; Cx ; Cy ; (M or m)
                self.read_unread_bytes();
                if !self.byte_buf.contains(&b'm') && !self.byte_buf.contains(&b'M') {
                    return Err(TuikitError::UnknownSequence(format!(
                        "ESC [ < (not ending with m/M)"
                    )));
                }

                let mut str_buf = String::new();
                let mut c = self.next_char_timeout(KEY_WAIT)?;
                while c != 'm' && c != 'M' {
                    str_buf.push(c);
                    c = self.next_char_timeout(KEY_WAIT)?;
                }
                let nums = &mut str_buf.split(';');

                let cb = nums.next().unwrap().parse::<u16>().unwrap();
                let cx = nums.next().unwrap().parse::<u16>().unwrap() - 1; // 0 based
                let cy = nums.next().unwrap().parse::<u16>().unwrap() - 1; // 0 based

                match cb {
                    0..=2 | 64..=65 => {
                        let button = match cb {
                            0 => MouseButton::Left,
                            1 => MouseButton::Middle,
                            2 => MouseButton::Right,
                            64 => MouseButton::WheelUp,
                            65 => MouseButton::WheelDown,
                            _ => {
                                return Err(TuikitError::UnknownSequence(format!(
                                    "ESC [ < {} {}",
                                    str_buf, c
                                )));
                            }
                        };

                        match c {
                            'M' => Ok(MousePress(button, cy, cx)),
                            'm' => Ok(MouseRelease(cy, cx)),
                            _ => Err(TuikitError::UnknownSequence(format!(
                                "ESC [ < {} {}",
                                str_buf, c
                            ))),
                        }
                    }
                    32 => Ok(MouseHold(cy, cx)),
                    _ => Err(TuikitError::UnknownSequence(format!(
                        "ESC [ < {} {}",
                        str_buf, c
                    ))),
                }
            }
            _ => Err(TuikitError::UnknownSequence(format!("ESC [ {:?}", seq2))),
        }
    }

    fn parse_cursor_report(&mut self) -> Result<Key> {
        self.read_unread_bytes();
        let pos_semi = self.byte_buf.iter().position(|&b| b == b';');
        let pos_r = self.byte_buf.iter().position(|&b| b == b'R');

        if pos_semi.is_some() && pos_r.is_some() {
            let pos_semi = pos_semi.unwrap();
            let pos_r = pos_r.unwrap();

            let remain = self.byte_buf.split_off(pos_r + 1);
            let mut col_str = self.byte_buf.split_off(pos_semi + 1);
            let mut row_str = std::mem::replace(&mut self.byte_buf, remain);

            row_str.pop(); // remove the ';' character
            col_str.pop(); // remove the 'R' character
            let row = String::from_utf8(row_str)?;
            let col = String::from_utf8(col_str)?;

            let row_num = row.parse::<u16>()?;
            let col_num = col.parse::<u16>()?;
            Ok(CursorPos(row_num - 1, col_num - 1))
        } else {
            Err(TuikitError::NoCursorReportResponse)
        }
    }

    fn extended_escape(&mut self, seq2: u8) -> Result<Key> {
        let seq3 = self.next_byte_timeout(KEY_WAIT)?;
        if seq3 == b'~' {
            match seq2 {
                b'1' | b'7' => Ok(Home), // tmux, xrvt
                b'2' => Ok(Insert),
                b'3' => Ok(Delete),     // kdch1
                b'4' | b'8' => Ok(End), // tmux, xrvt
                b'5' => Ok(PageUp),     // kpp
                b'6' => Ok(PageDown),   // knp
                _ => Err(TuikitError::UnknownSequence(format!("ESC [ {} ~", seq2))),
            }
        } else if seq3 >= b'0' && seq3 <= b'9' {
            let mut str_buf = String::new();
            str_buf.push(seq2 as char);
            str_buf.push(seq3 as char);

            let mut seq_last = self.next_byte_timeout(KEY_WAIT)?;
            while seq_last != b'M' && seq_last != b'~' {
                str_buf.push(seq_last as char);
                seq_last = self.next_byte_timeout(KEY_WAIT)?;
            }

            match seq_last {
                b'M' => {
                    // rxvt mouse encoding:
                    // ESC [ Cb ; Cx ; Cy ; M
                    let mut nums = str_buf.split(';');

                    let cb = nums.next().unwrap().parse::<u16>().unwrap();
                    let cx = nums.next().unwrap().parse::<u16>().unwrap() - 1; // 0 based
                    let cy = nums.next().unwrap().parse::<u16>().unwrap() - 1; // 0 based

                    match cb {
                        32 => Ok(MousePress(MouseButton::Left, cy, cx)),
                        33 => Ok(MousePress(MouseButton::Middle, cy, cx)),
                        34 => Ok(MousePress(MouseButton::Right, cy, cx)),
                        35 => Ok(MouseRelease(cy, cx)),
                        64 => Ok(MouseHold(cy, cx)),
                        96 | 97 => Ok(MousePress(MouseButton::WheelUp, cy, cx)),
                        _ => Err(TuikitError::UnknownSequence(format!("ESC [ {} M", str_buf))),
                    }
                }
                b'~' => {
                    let num: u8 = str_buf.parse().unwrap();
                    match num {
                        v @ 11..=15 => Ok(F(v - 10)),
                        v @ 17..=21 => Ok(F(v - 11)),
                        v @ 23..=24 => Ok(F(v - 12)),
                        200 => Ok(BracketedPasteStart),
                        201 => Ok(BracketedPasteEnd),
                        _ => Err(TuikitError::UnknownSequence(format!("ESC [ {} ~", str_buf))),
                    }
                }
                _ => unreachable!(),
            }
        } else if seq3 == b';' {
            let seq4 = self.next_byte_timeout(KEY_WAIT)?;
            if seq4 >= b'0' && seq4 <= b'9' {
                let seq5 = self.next_byte_timeout(KEY_WAIT)?;
                if seq2 == b'1' {
                    match (seq4, seq5) {
                        (b'5', b'A') => Ok(CtrlUp),
                        (b'5', b'B') => Ok(CtrlDown),
                        (b'5', b'C') => Ok(CtrlRight),
                        (b'5', b'D') => Ok(CtrlLeft),
                        (b'4', b'A') => Ok(AltShiftUp),
                        (b'4', b'B') => Ok(AltShiftDown),
                        (b'4', b'C') => Ok(AltShiftRight),
                        (b'4', b'D') => Ok(AltShiftLeft),
                        (b'3', b'H') => Ok(AltHome),
                        (b'3', b'F') => Ok(AltEnd),
                        (b'2', b'A') => Ok(ShiftUp),
                        (b'2', b'B') => Ok(ShiftDown),
                        (b'2', b'C') => Ok(ShiftRight),
                        (b'2', b'D') => Ok(ShiftLeft),
                        _ => Err(TuikitError::UnknownSequence(format!(
                            "ESC [ 1 ; {:x?} {:x?}",
                            seq4, seq5
                        ))),
                    }
                } else {
                    Err(TuikitError::UnknownSequence(format!(
                        "ESC [ {:x?} ; {:x?} {:x?}",
                        seq2, seq4, seq5
                    )))
                }
            } else {
                Err(TuikitError::UnknownSequence(format!(
                    "ESC [ {:x?} ; {:x?}",
                    seq2, seq4
                )))
            }
        } else {
            match (seq2, seq3) {
                (b'5', b'A') => Ok(CtrlUp),
                (b'5', b'B') => Ok(CtrlDown),
                (b'5', b'C') => Ok(CtrlRight),
                (b'5', b'D') => Ok(CtrlLeft),
                _ => Err(TuikitError::UnknownSequence(format!(
                    "ESC [ {:x?} {:x?}",
                    seq2, seq3
                ))),
            }
        }
    }

    // SSS3
    fn escape_o(&mut self) -> Result<Key> {
        let seq2 = self.next_byte_timeout(KEY_WAIT)?;
        match seq2 {
            b'A' => Ok(Up),    // kcuu1
            b'B' => Ok(Down),  // kcud1
            b'C' => Ok(Right), // kcuf1
            b'D' => Ok(Left),  // kcub1
            b'F' => Ok(End),   // kend
            b'H' => Ok(Home),  // khome
            b'P' => Ok(F(1)),  // kf1
            b'Q' => Ok(F(2)),  // kf2
            b'R' => Ok(F(3)),  // kf3
            b'S' => Ok(F(4)),  // kf4
            b'a' => Ok(CtrlUp),
            b'b' => Ok(CtrlDown),
            b'c' => Ok(CtrlRight), // rxvt
            b'd' => Ok(CtrlLeft),  // rxvt
            _ => Err(TuikitError::UnknownSequence(format!("ESC O {:x?}", seq2))),
        }
    }
}

pub struct KeyboardHandler {
    handler: Arc<SpinLock<File>>,
}

impl KeyboardHandler {
    pub fn interrupt(&self) {
        let mut handler = self.handler.lock();
        let _ = handler.write_all(b"x");
        let _ = handler.flush();
    }
}
